### ISSUE TITLE
CBLite ignoring channel removal without revision change

### ISSUE THREAD

**Issue by afeick** [couchbase/couchbase-lite-net#905]
[Archive.zip](https://github.com/couchbase/couchbase-lite-net/files/1304066/Archive.zip)

Using CBL 1.4.0
Couchbase Sync Gateway/1.4.1.2(4;56763dc)

The document is pushed up to the server, then the sync function removes the user's channel from that document. That change is sent down to the client in _changes feed for the same revision that the client already has, but there is no document change event and the document is not removed from the local database.

I've attached a Fiddler log with the two relevant requests and the CBLite.log file.

Here's the change response from the sync gateway with the user's channel in the removed array:

```
{
  "seq": "391135::391246",
  "id": "patientCareRecord_c0293a55-b082-e711-80ee-005056a8748b_bd83b490-b7be-42ce-8f56-b5b1ed23468f",
  "removed": [
    "tenant-product-user.c0293a55-b082-e711-80ee-005056a8748b.23fea179-bdea-e011-8887-005056c00008.793c374f-b082-e711-80ee-005056a8748b"
  ],
  "changes": [
    {
      "rev": "4-2a8fe1a469d6ff97ab5a38c620893cfd"
    }
  ]
}
```

**Comment by borrrden** [comment]
What do you mean by "without revision change"?  In a situation where a change in document A causes a removal of document B, Sync Gateway does not have a mechanism to send notification about the change in document B to the client and so the last known revision will remain until document B changes again.  Is this the situation you are talking about?

**Comment by afeick** [comment]
There is only one document. 

According to the documentation, the sync function has to call the [channel() function](https://developer.couchbase.com/documentation/mobile/current/guides/sync-gateway/sync-function-api-guide/index.html#channel-name) for all of the channels a revision should be routed to, otherwise the document will be removed from those channels. 

In the issue here, the sync function doesn't include the tenant-product-user... channel when it calls the channel() method.  Then the sync gateway sends a changed message to the client for that document telling it the user's channel was removed from the document (check the Fiddler archive above). Unfortunately CBLite seems to ignore this message.

**Comment by borrrden** [comment]
It looks like the user still has access to this document via another channel.  The client does realize the removed entry, as evidenced by the line "Removed entry received, body may not be available".  Nothing special happens on the client side at this time and instead it will go to sync gateway to get the next revision as normal.  If it has truly lost access then it will receive a special revision with the `_removed` property being set to `true` (or in rare cases where the document has been ejected by Couchbase Server for being too old, the client will get a 404 and skip it which is the reason for the line I mentioned above).  If not it will receive the next revision as normal.  Guest access can mess with this if it is on.  So can replicating against the admin port, but the Fiddler trace shows that the normal port is being used.  If you use the Sync Gateway REST API to query for this document and revision ID does it show the `_removed` property?

**Comment by afeick** [comment]
Here's the response if I try to get the document again with a session for that user:

> HTTP/1.1 403 Forbidden
> Content-Type: application/json
> Server: Couchbase Sync Gateway/1.4.1.2
> Date: Tue, 26 Sep 2017 14:22:51 GMT
> Content-Length: 42
> 
> {"error":"Forbidden","reason":"forbidden"}

**Comment by borrrden** [comment]
Hmmmm, could it be that there are conflicting revisions that are showing up once the current revision is removed?  Do you have a more complete Fiddler log that shows the _bulk_get request including the document in question (or a call to the document ID directly)?

**Comment by afeick** [comment]
Here's the rest of the session (other than login).  The removed message comes in the response for  # 39


[Removed Issue.saz.zip](https://github.com/couchbase/couchbase-lite-net/files/1337268/Removed.Issue.saz.zip)

**Comment by borrrden** [comment]
The session seems to indicate multiple simultaneous replications.  The logs also indicate at least three pullers and one pusher running.  Can you shed some light on this situation?  Perhaps another replication from a different user is also pulling into the same database?

**Comment by afeick** [comment]
There should be two pull replications and one push. We have two databases for each session. The first database is for user documents and it has a push and a pull replication. The other database is for shared, read-only configuration data and it only has a pull replication. (Our customers may have several users sharing the same device, and there are a number of large configuration files. Since they are usually on cellular connections we don't want to waste their data plan by replicating the same documents for every user.)

It looks like the CBLite.log I sent includes several sessions. The session in the fiddler log starts at Tue, 26 Sep 2017 21:38:18 GMT, which is 2017-09-26 15:38:19 (MDT)  in the logs . Starting from there I only see the three expected replications.

**Comment by borrrden** [comment]
Can you double check those date times?  I only have entries from Sept 14in the logs of either.

**Comment by afeick** [comment]
You're right, please ignore the last paragraph in that comment. I was looking at the logs for another issue that I just reported to the enterprise support site.  

In the CBLite log for this issue I still only see the expected two pullers and one pusher. There does appear to be duplicate log messages for everything, but that seems to be an issue with our log4net config unrelated to this issue.

```
2017-09-14 11:27:18,152 INFO  [TabletPCR.exe] [2924] [ZOLL.Mobile.Couchbase.CouchbaseLiteLogger] [SYNC (Replication)] Attempting to start puller (84dc41c1-c28f-4eb8-8675-87a68024ebd2)
...
2017-09-14 11:27:18,355 INFO  [TabletPCR.exe] [5124] [ZOLL.Mobile.Couchbase.CouchbaseLiteLogger] [SYNC (Replication)] Attempting to start puller (461f8ade-4660-4b66-9dfe-955af3e67070)
...
2017-09-14 11:27:18,621 INFO  [TabletPCR.exe] [2924] [ZOLL.Mobile.Couchbase.CouchbaseLiteLogger] [SYNC (Replication)] Attempting to start pusher (4f23289d-7191-4a1a-ac7a-4d07096ba139)
```

**Comment by borrrden** [comment]
I also see another one
> Attempting to start puller (0c828895-f31b-4de1-adc6-f03725e4075f)

which occurs about 1 second later than those three.

**Comment by borrrden** [comment]
Also in the message above where you show the 403, is that requesting the doc via doc ID and revision ID, or just doc ID?

i.e.
`docname?rev=4-123456abcdef/`

or

`docname/`

**Comment by borrrden** [comment]
Oddly on the "removed issue" fiddler trace I don't see any requests for new documents after the request for changes.  Either they didn't happen, or the client decided that it already had all of the revisions that it received.  One word of warning is that if the client already has the full version of the revision that was removed it won't be overwritten by a subsequent removed entry.

**Comment by afeick** [comment]
> One word of warning is that if the client already has the full version of the revision that was removed it won't be overwritten by a subsequent removed entry (that is, if it has revision 5, and then it gets revision 5 removed it won't overwrite revision 5)

This is likely the issue then.  The sync function is removing the document from a channel without updating the revision, so then the client isn't removing it. The [documentation for the channel() function](https://developer.couchbase.com/documentation/mobile/current/guides/sync-gateway/sync-function-api-guide/index.html#channel-name) is a little confusing then because it says the client "will receive an empty revision of the document with nothing but a "_removed": true property"

Also, you were right about the third puller. We are creating two pullers for the same database, but they're subscribed to different channels. (My apologies for the confusion, it looks like someone on my team changed that since I touched that code.) 

Do you know if there are any issues with multiple pullers on the same database?

**Comment by borrrden** [comment]
> Do you know if there are any issues with multiple pullers on the same database?

Not in general, but it can lead to unexpected results.  One that I can think of that is relevant to this case is that one puller will pull down the actual revision, while the other gets a removed entry.  This is a race since these two revisions will have the same ID.  Revisions are immutable and replication is append-only so once you have one of them in the local database that is the one you are stuck with.  A removed entry will be inserted just as any other normal revision, but just as any other normal revision if the local database already has a revision with the same ID then it will ignore that entry on the changes feed.  

I don't think that is what is happening here.  I think what you've run into is when a client tries to remove access from itself.  If I recall correctly, that is not going to work because the client itself made the change locally and so it already has to whole revision.  It will stop receiving updates after that though since it no longer has access.  If *another* client were to make the same change you would see the document immediately disappear.  This is because of the immutability and append-only characteristics mentioned above.  

@snej I noticed a [bug report](https://github.com/couchbase/couchbase-lite-ios/issues/671) from about 2 years ago discussing this.  The issue is different but there is some discussion inside and the same thing looks like it is happening.  Do you remember if this was addressed and .NET is doing things wrong?

**Comment by borrrden** [comment]
Ok I discussed this situation with the mobile architect and though I'm guessing this will not be an answer you like this is how things need to work.  Consider this:

A database can replicate (push) to any number of other endpoints and so it needs to store a copy of what to push so that it can continue to do so.  That's the reason for the tenant of revision immutability.  To accomplish the above would require a mutation to an immutable object, which means a race will be created.  The pull could alter the revision in between two pushes and then you'd end up with inconsistent data (this may seem like inconsistent data but it is more accurately inconsistent access).  If you know your sync function, then perhaps you could set up a view which will map the documents that you need.  If you are worried about sensitive information then after you run a push to ensure the revision has been pushed, you could purge the revision from the local database.

So in simple terms, you can't lose access to a revision that you created locally.

**Comment by djpongh** [comment]
Working as designed.
